gpkit/__init__.py

```diff
@@ -115,6 +116,59 @@ def __call__(self, arg):
 
 enable_units()
 
+VECTORIZATION = []
+MODELS = []
+MODELNUMS = []
+MODELNUM_LOOKUP = defaultdict(int)
+NAMEDVARS = defaultdict(list)
+
+def begin_variable_naming(model):
+    "Appends a model name and num to the environment."
+    MODELS.append(model)
+    num = MODELNUM_LOOKUP[model]
+    MODELNUMS.append(num)
+    MODELNUM_LOOKUP[model] += 1
+    return num, (tuple(MODELS), tuple(MODELNUMS))
+
+
+def end_variable_naming():
+    "Pops a model name and num from the environment."
+    NAMEDVARS.pop(tuple(MODELS), tuple(MODELNUMS))
+    MODELS.pop()
+    MODELNUMS.pop()
+
+
+class NamedVariables(object):
+    """Creates an environment in which all variables have
+       a model name and num appended to their varkeys.
+    """
+    def __init__(self, model):
+        self.model = model
+
+    def __enter__(self):
+        "Enters a named environment."
+        begin_variable_naming(self.model)
+
+    def __exit__(self, type_, val, traceback):
+        "Leaves a named environment."
+        end_variable_naming()
+
+
+class Vectorize(object):
+    """Creates an environment in which all variables are
+       exended in an additional dimension.
+    """
+    def __init__(self, dimension_length):
+        self.dimension_length = dimension_length
+
+    def __enter__(self):
+        "Enters a vectorized environment."
+        VECTORIZATION.insert(0, self.dimension_length)
+
+    def __exit__(self, type_, val, traceback):
+        "Leaves a vectorized environment."
+        VECTORIZATION.pop(0)
+
```

Comment:
Yeah, it would! But importing a python linked-list implementation that rolls the other way seems a bit silly...